### PR TITLE
fix: add logic to handle fixTypes in the lintText()

### PR DESCRIPTION
**Pull request by amareshsm**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)
fixes https://github.com/eslint/eslint/issues/18732

### CODE REVIEW

**Comment by fasttime** [comment]
Thanks for the PR! I think we need unit tests for this fix. The tests we currently have for `fixTypes` only call `lintFiles()`: https://github.com/eslint/eslint/blob/v9.8.0/tests/lib/eslint/eslint.js#L5114-L5170. We could simply add similar tests that call `lintText()`.

**Comment by amareshsm** [comment]
> Thanks for the PR! I think we need unit tests for this fix. The tests we currently have for `fixTypes` only call `lintFiles()`: https://github.com/eslint/eslint/blob/v9.8.0/tests/lib/eslint/eslint.js#L5114-L5170. We could simply add similar tests that call `lintText()`.

👍 I will add Unit tests.

**Comment by aladdin-add** [comment]
it needs some tests, so converting to a draft to avoid it being merged by accident.

**Comment by amareshsm** [comment]
UTs added for lintText() also.

File: tests/lib/eslint/eslint.js [hunk 1]
```diff
@@ -5100,73 +5100,155 @@ describe("ESLint", () => {
...
+                        cwd: path.join(fixtureDir, ".."),
+                        overrideConfigFile: true,
+                        fix: true,
+                        fixTypes: ["layout"]
+                    });
+                    const inputPath = getFixturePath("fix-types/fix-only-semi.js");
+
+                    await eslint.lintText([inputPath]);
+                }, /'code' must be a string\./iu);
             });
```
Comment (fasttime) [inline tests/lib/eslint/eslint.js:5189]: This is an interesting test case, but my feeling is that it does not belong in this test suite (you'd get the same error without `fix` or `fixTypes`). A similar, but more minimal test case with a non-string argument is already included in the "lintText()" suite:

https://github.com/eslint/eslint/blob/41d02066935b987d2e3b13a08680cc74d7067986/tests/lib/eslint/eslint.js#L905-L908

File: tests/lib/eslint/eslint.js [hunk 2]
```diff
@@ -5100,73 +5100,155 @@ describe("ESLint", () => {
...
-        });
+            it("should not fix any rules when fixTypes is used without fix", async () => {
+                eslint = new ESLint({
+                    cwd: path.join(fixtureDir, ".."),
+                    overrideConfigFile: true,
+                    fix: false,
+                    fixTypes: ["layout"]
+                });
+                const inputPath = getFixturePath("fix-types/fix-only-semi.js");
+                const content = await fs.readFileSync(inputPath, "utf8");
```
Comment (fasttime) [inline tests/lib/eslint/eslint.js:5199]: `await` is not required for synchronous calls like `fs.readFileSync`. But it is required for asynchronous calls like `fsp.readFile` (from Node.js [fs Promises API](https://nodejs.org/dist/latest-v10.x/docs/api/fs.html#fs_fs_promises_api)).

So we could use one of `fs.readFileSync` (without `await`) or `await fsp.readFile` here. Whichever you choose, I would just suggest to stay consistent and use the same call in all the other tests, as that makes the code clearer.

File: tests/lib/eslint/eslint.js (hunk 1 above)
Comment (amareshsm) [reply tests/lib/eslint/eslint.js:5189]: I will remove this test.

File: tests/lib/eslint/eslint.js (hunk 2 above)
Comment (amareshsm) [reply tests/lib/eslint/eslint.js:5199]: yeah my bad I was previously using `fsp.readFil`e and switched to `fs.readFileSync` but forgot to remove the await. Thanks!

**Review by fasttime — COMMENTED** [review]
Thanks! This is looking good already. I just left a suggestion to remove a duplication.

File: lib/eslint/eslint.js [hunk 3]
```diff
@@ -1032,14 +1032,23 @@ class ESLint {
...
         const results = [];
         const startTime = Date.now();
+        const fixTypesSet = fixTypes ? new Set(fixTypes) : null;
+        let fixer = fix;
         const resolvedFilename = path.resolve(cwd, filePath || "__placeholder__.js");
 
+        if (fix && fixTypesSet) {
+            const originalFix = (typeof fix === "function") ? fix : () => true;
+
+            fixer = message => shouldMessageBeFixed(message, configs.getConfig(resolvedFilename), fixTypesSet) && originalFix(message);
```
[the remark this answers is not in the document]
Comment (amareshsm) [reply lib/eslint/eslint.js]: done

File: lib/eslint/eslint.js [hunk 4]
```diff
@@ -570,6 +570,23 @@ function createExtraneousResultsError() {
     return new TypeError("Results object was not created from this ESLint instance.");
 }
 
+/**
+ * Creates a fixer function based on the provided fix, fixTypesSet, and filePath.
```
Comment (fasttime) [inline lib/eslint/eslint.js:574]:
```suggestion
 * Creates a fixer function based on the provided fix, fixTypesSet, and config.
```

Comment (amareshsm) [reply lib/eslint/eslint.js]: done

**Review by fasttime — APPROVED** [review]
LGTM, thanks!

**Pull request merged**